sstables/mirrored_storage.cc

```diff
@@ -0,0 +1,352 @@
+    co_return data_sink{std::make_unique<write_through_sink>(std::move(local_sink),
+                                                             std::move(remote_sink))};
+}
+
+future<data_sink> mirrored_storage::make_component_sink(sstable& sst,
+                                                      component_type type,
+                                                      open_flags oflags,
+                                                      file_output_stream_options options) {
+    auto s3_object_name = _make_s3_object_name(sst, type);
+    auto remote_sink = _s3_client->make_upload_sink(s3_object_name);
+
+    auto fd = co_await open_component(sst, type, oflags, {}, false);
+    auto local_sink = co_await make_file_data_sink(std::move(fd), options);
+    // in general, the sizes of these components are small. so always
+    // update the local cache in "write_through" mode
+    auto sink_impl = std::make_unique<write_through_sink>(std::move(local_sink),
+                                                          std::move(remote_sink));
+    co_return data_sink{std::move(sink_impl)};
+}
+
+future<> mirrored_storage::destroy(const sstable& sst) {
+    return make_ready_future();
+}
+
+static future<> delete_sstables(std::vector<shared_sstable> ssts) {
+    // TODO: delete the sstables with a transaction semantics
+    co_await coroutine::parallel_for_each(ssts, [] (shared_sstable sst) -> future<> {
+        return sst->unlink();
+    });
+}
+
+noncopyable_function<future<>(std::vector<shared_sstable>)> mirrored_storage::atomic_deleter() const {
+    return delete_sstables;
+}
+
+future<> mirrored_storage::remove_by_registry_entry(entry_descriptor desc) {
+    // used when performing garbage_collect as requested by sstable_directory::prepare()
+    // system_keyspace_components_lister uses this to remove the components and the
+    // systable entries yet marked "sealed".
+    auto prefix = format("/{}/{}", _bucket, desc.generation);
+    std::optional<temporary_buffer<char>> toc;
+    std::vector<sstring> components;
+
+    try {
+        toc = co_await _s3_client->get_object_contiguous(prefix + "/" + sstable_version_constants::get_component_map(desc.version).at(component_type::TOC));
+    } catch (seastar::httpd::unexpected_status_error& e) {
+        if (e.status() != seastar::http::reply::status_type::no_content) {
+            throw;
+        }
+    }
+    if (!toc) {
+        co_return; // missing TOC object is OK
+    }
+
+    boost::split(components, std::string_view(toc->get(), toc->size()), boost::is_any_of("\n"));
+    co_await coroutine::parallel_for_each(components, [this, &prefix] (sstring comp) -> future<> {
+        if (comp != sstable_version_constants::TOC_SUFFIX) {
+            co_await _s3_client->delete_object(prefix + "/" + comp);
+        }
+    });
+    co_await _s3_client->delete_object(prefix + "/" + sstable_version_constants::TOC_SUFFIX);
```

Comment:
It looks like it only garbage collects the remote part of the storage, but how about local files?